File: patterns-use-cases/async-signals-payment/async-signals-payment-typescript/src/utils/stripe_utils.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */

import { TerminalError } from "@restatedev/restate-sdk";
import Stripe from "stripe";

const stripeSecretKey = "...";
const webHookSecret = "...";

const stripe = new Stripe(stripeSecretKey, { apiVersion: "2023-10-16" });

export const RESTATE_CALLBACK_ID = "restate_callback_id";

export function parseWebhookCall(requestBody: any, signature: string | string[]) {
    return stripe.webhooks.constructEvent(requestBody, signature, webHookSecret);
}

export async function createPaymentIntent(request: {
    paymentMethodId: string,
    amount: number,
    idempotencyKey: string,
    webhookPromiseId: string,
    delayedStatus?: boolean
}): Promise<Stripe.PaymentIntent> {

    const requestOptions = {
        idempotencyKey: request.idempotencyKey
    }

    try {
        const paymentIntent: Stripe.PaymentIntent = await stripe.paymentIntents.create({
            amount: request.amount,
            currency: "usd",
            payment_method: request.paymentMethodId,
            confirm: true,
            confirmation_method: "automatic",
            return_url: "https://restate.dev/", // some random URL
            metadata: {
                restate_callback_id: request.webhookPromiseId
            }
        }, requestOptions);

        // simulate delayed notifications for testing
        if (request.delayedStatus) {
            paymentIntent.status = "processing";
        }
        return paymentIntent;
    }
    catch (error) {
        if (error instanceof Stripe.errors.StripeCardError) {
            // simulate delayed notifications for testing
            const paymentIntent = error.payment_intent;
            if (request.delayedStatus && paymentIntent) {
                paymentIntent.status = "processing";
                return paymentIntent;
             } else {
                throw new TerminalError(`Payment declined: ${paymentIntent?.status} - ${error.message}`);
             }
        } else {
            throw error;
        }
    }
}

File: patterns-use-cases/async-signals-payment/async-signals-payment-java/src/main/java/my/example/utils/StripeUtils.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */
package my.example.utils;

import com.stripe.StripeClient;
import com.stripe.exception.SignatureVerificationException;
import com.stripe.exception.StripeException;
import com.stripe.model.Event;
import com.stripe.model.EventDataObjectDeserializer;
import com.stripe.model.PaymentIntent;
import com.stripe.net.RequestOptions;
import com.stripe.net.Webhook;
import com.stripe.param.PaymentIntentCreateParams;
import com.stripe.param.PaymentIntentCreateParams.ConfirmationMethod;
import dev.restate.sdk.common.TerminalException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class StripeUtils {

  Logger logger = LogManager.getLogger(StripeUtils.class);
  private final String stripeSecretKey = "sk_test_...";
  private final String webhookSecret = "whsec_...";
  private final StripeClient stripe;

  public StripeUtils() {
    stripe = StripeClient.builder().setApiKey(stripeSecretKey).build();
  }

  public PaymentIntent createPaymentIntent(
      String paymentMethodId,
      Long amount,
      String idempotencyKey,
      String webhookPromiseId,
      boolean delayedStatus) {

    try {
      PaymentIntent paymentIntent =
          stripe
              .paymentIntents()
              .create(
                  new PaymentIntentCreateParams.Builder()
                      .setPaymentMethod(paymentMethodId)
                      .setAmount(amount)
                      .setCurrency("USD")
                      .setConfirm(true)
                      .setConfirmationMethod(ConfirmationMethod.AUTOMATIC)
                      .setReturnUrl("https://restate.dev/")
                      .putMetadata("restate_callback_id", webhookPromiseId)
                      .build(),
                  RequestOptions.builder().setIdempotencyKey(idempotencyKey).build());

      if (delayedStatus) {
        paymentIntent.setStatus("processing");
      }

      return paymentIntent;
    } catch (StripeException err) {
      logger.error("Payment error: " + err.getMessage());
      // Simulate delayed notifications for testing
      try {
        PaymentIntent paymentIntent = err.getStripeError().getPaymentIntent();
        if (delayedStatus) {
          paymentIntent.setStatus("processing");
          return paymentIntent;
        } else {
          throw new TerminalException(
              "Payment declined: " + paymentIntent.getStatus() + " - " + err.getMessage());
        }
      } catch (NullPointerException exc) {
        throw new TerminalException("Payment error: " + exc.getMessage());
      }
    }
  }

  public Event parseWebhookCall(byte[] request, String sig) {
    try {
      return Webhook.constructEvent(new String(request), sig, webhookSecret);
    } catch (SignatureVerificationException e) {
      throw new TerminalException(400, "Invalid Stripe signature");
    }
  }

  public PaymentIntent parseAsPaymentIntent(Event event) {
    EventDataObjectDeserializer dataObjectDeserializer = event.getDataObjectDeserializer();
    PaymentIntent paymentIntent = null;
    if (dataObjectDeserializer.getObject().isPresent()) {
      paymentIntent = (PaymentIntent) dataObjectDeserializer.getObject().get();
    } else {
      throw new TerminalException(500, "No Stripe object found in event");
    }

    return paymentIntent;
  }
}
